packages/flutter/lib/src/widgets/app.dart

```diff
@@ -82,19 +82,18 @@ class WidgetsApp extends StatefulWidget {
   ///  _not_ contain an entry for `'/'`.  Conversely, if [home] is omitted, [routes]
   /// _must_ contain an entry for `'/'`.
   ///
-  /// If [home] or [routes] are not null, then either the [pageRoutebuilder] or
-  /// the [builder] parameter is required. These parameters will be used so
-  /// that the default routing implementation in [WidgetsApp] can wrap routes in
-  /// appropriate transitions. For example, [MaterialApp] will provide a
-  /// [pageRoutebuilder] that creates Material compliant hero animations between
-  /// routes, whereas the [CupertinoApp] provides Cupertino compliant hero
-  /// animations. Other implementations can provide other custom transitions here.
-  ///
-  /// The [builder] parameter is optional in all cases. It can be used to ensure that
-  /// all route entries get wrapped in another widget. It is invoked during the build
-  /// phase of this widget.  If it is specified,
-  ///
-  /// It is also possible to provide a custom implementation of routing via the
+  /// If [home] or [routes] are not null, the routing implementation needs to know how
+  /// appropriately build [PageRoutes]. This can be achieved by supplying the 
+  /// [pageRouteBuilder] parameter or the [builder] parameter, or both.  The 
+  /// [pageRouteBuilder] is more appropriate if you will only be building [Route]
```

Comment:
They're also not really at the same level of decision no? The pageRouteBuilder is fed into the _onGenerateRoute when building individual routes. The builder is outside the route and outside the navigator altogether.

Reply:
As implemented, pageRouteBuilder will get ignored if `builder` is specified - that may be wrong, but it's how it is.  Maybe it's be more appropriate to say `the routing implementation needs to know how to appropriately build a new widget when routing`?

Reply:
Sounds unexpected. Perhaps you're referring to https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/widgets/app.dart#L646? (in that case we should probably rename the similarly named local variable). 

I also meant we should clarify what the choice the user is to make (which are at different scales of 'complexity'). If you override `pageRouteBuilder`, given the page content builder that the user already provided, decorate it with a route. If you override `builder`, you're building everything. That difference in scale should be described.

Reply:
Ahh you are absolutely right.  I'll update the doc and rename the local there.

Reply:
Overriding builder doens't really mean overriding everything though - builder gets passed a child, it's meant to just be a wrapper around whatever we'd be doing normally.